runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPCollectionFusers.java

```diff
@@ -181,6 +181,10 @@ private static boolean parDoCompatibility(
     // side inputs can be fused with other transforms in the same environment which are not
     // upstream of any of the side inputs.
     return pipeline.getSideInputs(parDo).isEmpty()
+        // Since we lack the ability to mark upstream transforms as key preserving, we
+        // purposefully break fusion here to provide runners the opportunity to insert a
+        // grouping operation
+        && pipeline.getUserStates(parDo).isEmpty()
```

Comment:
Based on what information would a runner insert a shuffle though? Unless that is available, there is also no benefit breaking the chain here?

Reply:
Runners need to look at the state spec and see if it exists, that would signal insertion of anything which does a partition by key (e.g. a reshuffle). Most runners (maybe all) only support user state via key partitioning.